src/data_generators/copula_data_set.py

```diff
@@ -0,0 +1,176 @@
+    DataSet,
+)
+from wfa_planning_evaluation_framework.data_generators.pricing_generator import (
+    PricingGenerator,
+)
+from wfa_planning_evaluation_framework.data_generators.fixed_price_generator import (
+    FixedPriceGenerator,
+)
+
+
+class CopulaDataSet(DataSet):
+    """A DataSet constructed by gluing a list of PublisherData with Copula."""
+
+    def __init__(
+        self,
+        unlabeled_publisher_data_list: Iterable[PublisherData],
+        copula_generator: Copula = GaussianCopula(corr=0),
+        universe_size: int = None,
+        pricing_generator: PricingGenerator = FixedPriceGenerator(0.1),
+        random_generator: np.random.Generator = np.random.default_rng(0),
+        name: str = "copula",
+    ):
+        """Constructor for OverlapDataSet.
+
+        Args:
+            unlabeled_publisher_data_list:  A list of PublisherData. Each PublisherData
+                indicates the frequency distribution at a publisher --- the
+                ids labels are meaningless while their distribution matters.
+            copula_generator:  An instance of a subclass of
+                statsmodels.distributions.copula.copulas.Copula.  See here for the
+                possible choices:
+                https://github.com/statsmodels/statsmodels/tree/32dc52699f994acbf9dbdb9bd10d7eff04d860f5/statsmodels/distributions
+            universe_size:  A cross-publisher universe size to construct the copula.
+                If not given, then specify the universe size as the maximum single
+                publisher reach times 2.
+            pricing_generator:  A PricingGenerator object that annotates a list of
+                id's with randomly generated price information.  At this moment,
+                assume that all the publishers share the same PricingGenerator.
+            random_generator:  A random generator to draw sample from the copula.
+            name:  If specified, a human-readable name that will be associated to this
+                DataSet.
+        """
+        if universe_size is None:
+            universe_size = 2 * max(
+                [1] + [data.max_reach for data in unlabeled_publisher_data_list]
+            )
+        self.marginal_distributions = []
+        for pub in unlabeled_publisher_data_list:
+            pmf = self.zero_included_pmf(pub, universe_size)
+            self.marginal_distributions.append(
+                stats.rv_discrete(values=(range(len(pmf)), pmf))
+            )
+        self.copula_generator = copula_generator
+        self.distribution = CopulaDistribution(
+            copula=self.copula_generator,
+            marginals=self.marginal_distributions,
+        )
+        self.sample = self.distribution.rvs(
+            nobs=universe_size,
+            random_state=random_generator.integers(0, 1e9),
+        ).astype("int32")
+        imps = self.to_impressions(self.sample)
+        for pub_imps in imps:
```

Comment:
suggestion: The shuffling that Matthew suggested can also be done inside or the `to_impression` method

Reply:
Makes sense.  Now moved the shuffling to `to_impressions`.